### ISSUE TITLE
Improper handling of "unusual" enums in RuntimeType

### ISSUE THREAD

**Issue by matty-hall** [dotnet/coreclr#1376]
RuntimeType.GetEnumName and RunTimeType.IsEnumDefined both throw exceptions for enums whose underlying type is IntPtr or UIntPtr.  I haven't tested other code paths extensively, but I'd imagine that there are probably other places that this issue surfaces due to how unusual this construct is.  Things actually do work for enum types with an underlying type of Boolean or Char, which is somewhat interesting given that those feel less "integery" to me than native integers (even though clearly they're ints on the stack).

Now, while C# doesn't allow you to create such an enum, the spec does (which is, I presume, why peverify has no problems with this construct either).  Here's the relevant portion of the spec:

> I.11 Collected Common Language Specification rules
> **CLS Rule 7**: The underlying type of an enum shall be a built -in CLS integer type, the name of the field shall be "value__", and that field shall be marked RTSpecialName. (§I.8.5.2)

That's not very specific, but elsewhere it is:

> II.14.3 Enums
> An enum (short for enumeration) defines a set of symbols that all have the same type. A type shall be an enum if and only if it has an immediate base type of System.Enum. Since System.Enum itself has an immediate base type of System.ValueType, (see Partition IV) enums are value types (§II.13) The symbols of an enum are represented by an underlying integer type: one of { **bool, char, int8, unsigned int8, int16, unsigned int16, int32, unsigned int32, int64, unsigned int64, native int, unsigned native int** }

So, while I get that this is odd, the spec certainly allows it, and it seems arbitrary that bool and char work but IntPtr/UIntPtr don't.

Repro:

```
.assembly extern mscorlib
{
    .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )
    .ver 4:0:0:0
}
.assembly EnumTest
{
    .hash algorithm 0x00008004
    .ver 1:0:0:0
}

.module EnumTest.exe
.imagebase 0x00400000
.file alignment 0x00000200
.stackreserve 0x00100000
.subsystem 0x0003
.corflags 0x00020003

.class private auto ansi beforefieldinit EnumTest.Program extends [mscorlib]System.Object
{
    .method private hidebysig static void  Main(string[] args) cil managed
    {
        .entrypoint

        ldstr      "   BoolEnum: Value={0} Validity={1}"
        ldc.i4.1
        call       string EnumTest.Program::EnumToString<valuetype EnumTest.BooleanEnum,bool>(!!1)
        ldc.i4.1
        call       string EnumTest.Program::EnumValidity<valuetype EnumTest.BooleanEnum,bool>(!!1)
        call       void [mscorlib]System.Console::WriteLine(string, object, object)

        ldstr      "   CharEnum: Value={0} Validity={1}"
        ldc.i4.s   65
        call       string EnumTest.Program::EnumToString<valuetype EnumTest.CharEnum,char>(!!1)
        ldc.i4.s   65
        call       string EnumTest.Program::EnumValidity<valuetype EnumTest.CharEnum,char>(!!1)
        call       void [mscorlib]System.Console::WriteLine(string, object, object)

        ldstr      " IntPtrEnum: Value={0} Validity={1}"
        ldc.i4.1
        newobj     instance void [mscorlib]System.IntPtr::.ctor(int32)
        call       string EnumTest.Program::EnumToString<valuetype EnumTest.IntPtrEnum,native int>(!!1)
        ldc.i4.1
        newobj     instance void [mscorlib]System.IntPtr::.ctor(int32)
        call       string EnumTest.Program::EnumValidity<valuetype EnumTest.IntPtrEnum,native int>(!!1)
        call       void [mscorlib]System.Console::WriteLine(string, object, object)

        ldstr      "UIntPtrEnum: Value={0} Validity={1}"
        ldc.i4.2
        newobj     instance void [mscorlib]System.UIntPtr::.ctor(uint32)
        call       string EnumTest.Program::EnumToString<valuetype EnumTest.UIntPtrEnum,native uint>(!!1)
        ldc.i4.2
        newobj     instance void [mscorlib]System.UIntPtr::.ctor(uint32)
        call       string EnumTest.Program::EnumValidity<valuetype EnumTest.UIntPtrEnum,native uint>(!!1)
        call       void [mscorlib]System.Console::WriteLine(string, object, object)

        ret
    }

    .method public hidebysig static string EnumToString<E,U>(!!U underlyingValue) cil managed
    {
        .locals init (
            [0] object boxed,
            [1] !!E enumValue,
            [2] class [mscorlib]System.Exception ex,
            [3] string CS$1$0000
        )

        ldarg.0
        box        !!U
        stloc.0
        ldloc.0
        unbox.any  !!E
        stloc.1

        .try
        {
            ldloca.s   enumValue
            constrained. !!E
            callvirt   instance string [mscorlib]System.Object::ToString()
            stloc.3
            leave.s    afterTry
        }
        catch [mscorlib]System.Exception 
        {
            stloc.2
            ldstr      "[exception: "
            ldloc.2
            callvirt   instance string [mscorlib]System.Exception::get_Message()
            ldstr      "]"
            call       string [mscorlib]System.String::Concat(string,string,string)
            stloc.3
            leave.s    afterTry
        }

    afterTry:
        ldloc.3
        ret
    }

    .method public hidebysig static string EnumValidity<E,U>(!!U underlyingValue) cil managed
    {
        .locals init (
            [0] object boxed,
            [1] class [mscorlib]System.Exception ex,
            [2] string CS$1$0000
        )

        ldarg.0
        box        !!U
        stloc.0

        .try
        {
            ldtoken    !!E
            call       class [mscorlib]System.Type [mscorlib]System.Type::GetTypeFromHandle(valuetype [mscorlib]System.RuntimeTypeHandle)
            ldloc.0
            call       bool [mscorlib]System.Enum::IsDefined(class [mscorlib]System.Type,object)
            brfalse.s  enumInvalid

            ldstr      "valid"
            stloc.2
            leave.s    afterTry

      enumInvalid:
            ldstr      "invalid"
            stloc.2
            leave.s    afterTry
        }
        catch [mscorlib]System.Exception
        {
              stloc.1
              ldstr      "[exception: "
              ldloc.1
              callvirt   instance string [mscorlib]System.Exception::get_Message()
              ldstr      "]"
              call       string [mscorlib]System.String::Concat(string,string,string)
              stloc.2
              leave.s    afterTry
        }

    afterTry:
        ldloc.2
        ret
    }

    .method public hidebysig specialname rtspecialname instance void  .ctor() cil managed
    {
        ldarg.0
        call       instance void [mscorlib]System.Object::.ctor()
        ret
    }
}

.class public auto ansi sealed EnumTest.BooleanEnum extends [mscorlib]System.Enum
{
    .field public specialname rtspecialname bool value__
    .field public static literal valuetype EnumTest.BooleanEnum Yes = bool(true)
}

.class public auto ansi sealed EnumTest.CharEnum extends [mscorlib]System.Enum
{
    .field public specialname rtspecialname char value__
    .field public static literal valuetype EnumTest.CharEnum TheLetterA = char(65)
}

.class public auto ansi sealed EnumTest.IntPtrEnum extends [mscorlib]System.Enum
{
    .field public specialname rtspecialname native int value__
    .field public static literal valuetype EnumTest.IntPtrEnum One = uint32(0x01)
}

.class public auto ansi sealed EnumTest.UIntPtrEnum extends [mscorlib]System.Enum
{
    .field public specialname rtspecialname native unsigned int value__
    .field public static literal valuetype EnumTest.UIntPtrEnum Two = uint32(0x02)
}
```

**Comment by matty-hall** [comment]
Small update here -- I found another issue here when you create a custom attribute that contains a property/field of an enum type whose underlying type is IntPtr or UIntPtr (things work fine for bool/char enums).  The CLR crashes when a call to GetCustomAttributes tries to fetch such an attribute.

The behavior leads me to believe that the text I quoted above about support for IntPtr/UIntPtr enums is actually a bug in the ECMA specification.  Specifically, section II.23.3 (Custom Attributes) doesn't mention IntPtr/UIntPtr values at all in the description of enums in attribute serialization, nor does section II.22.9 (Constants).  Enums whose underlying type is Char/Bool seem to work if only because there's a clear way to serialize them (as ELEMENT_TYPE_BOOLEAN or ELEMENT_TYPE_CHAR).  One would think that if the ECMA spec really intended to support IntPtr/UIntPtr enums, then it would have specified this behavior for them.
